antelope_contracts/contracts/erc20/include/erc20/erc20.hpp

```diff
@@ -11,29 +11,116 @@ using namespace intx;
+    uint8_t buffer[32]={0};
+    check(len <= sizeof(buffer), "invalida size");
+    memcpy(buffer, ptr, len);
+    return checksum256(buffer);
+}
+
+checksum256 make_key(bytes data){
+    return make_key((const uint8_t*)data.data(), data.size());
+}
+
 class [[eosio::contract]] erc20 : public contract {
    public:
     using contract::contract;
 
+   struct bridge_message_v0 {
+      eosio::name        receiver;
+      bytes              sender;
+      eosio::time_point  timestamp;
+      bytes              value;
+      bytes              data;
+
+      EOSLIB_SERIALIZE(bridge_message_v0, (receiver)(sender)(timestamp)(value)(data));
+   };
+
+   using bridge_message_t = std::variant<bridge_message_v0>;
+
     [[eosio::on_notify("*::transfer")]] void transfer(eosio::name from, eosio::name to, eosio::asset quantity, std::string memo);
 
     // evm runtime will call this to notify erc20 about the message from 'from' with 'data'.
-    [[eosio::action]] void onbridgemsg(name receiver, const bytes& sender, const time_point& timestamp, const bytes& value, const bytes& data);
+    [[eosio::action]] void onbridgemsg(const bridge_message_t &message);
     [[eosio::action]] void init();
 
-    struct [[eosio::table]] config
-   {
-      bytes erc20_addr;
+    [[eosio::action]] void regtoken(eosio::name eos_contract_name, 
+    std::string evm_token_name, std::string evm_token_symbol, const eosio::asset& min_deposit, const eosio::asset& deposit_fee, const eosio::asset &egress_fee, std::string erc20_impl_address, int erc20_precision);
+
+    [[eosio::action]] void addegress(const std::vector<name>& accounts);
+    [[eosio::action]] void removeegress(const std::vector<name>& accounts);
+
+   uint64_t my_next_nonce();
+
+   struct [[eosio::table("nextnonces")]] nextnonce {
+      name     owner;
+      uint64_t next_nonce = 0;
 
-      EOSLIB_SERIALIZE(config, (erc20_addr));
+      uint64_t primary_key() const { return owner.value; }
+
+      EOSLIB_SERIALIZE(nextnonce, (owner)(next_nonce));
+   };
+   typedef eosio::multi_index<"nextnonces"_n, nextnonce> nextnonces_table_t;
+
+   struct [[eosio::table("implcontract")]] impl_contract_t {
+      uint64_t       id = 0;
+      bytes          address;
+
+      uint64_t       primary_key() const {
+         return id;
+      }
+      checksum256    by_address()const { 
+        return make_key(address);
+      }
+   };
+   typedef eosio::multi_index<"implcontract"_n, impl_contract_t,
+      indexed_by<"by.address"_n, const_mem_fun<impl_contract_t,  checksum256, &impl_contract_t::by_address> > 
+      > impl_contract_table_t;
+
+   struct [[eosio::table("tokens")]] token_t {
+      uint64_t       id = 0;
+      eosio::name    eos_contract_name; 
+      bytes          address; // <-- proxy contract addr
+      eosio::asset   min_deposit;
+      eosio::asset   deposit_fee;
+      uint64_t       balance = 0; // <-- total amount in EVM side, using native side precision
```

Comment:
Instead of including the raw `uint64_t`, it would be more friendly to make this an `eosio::asset`. Yes it would include a redundant 8 bytes for the symbol, but it would be friendly to view in block explorers.

Reply:
fixed